diesel/src/pg/types/numeric.rs

```diff
@@ -5,91 +5,62 @@ mod bigdecimal {
     extern crate num_integer;
     extern crate bigdecimal;
 
+    use self::bigdecimal::BigDecimal;
+    use self::num_bigint::{Sign, BigInt, BigUint};
+    use self::num_integer::Integer;
+    use self::num_traits::{Signed, Zero, ToPrimitive};
     use std::error::Error;
     use std::io::prelude::*;
 
     use pg::Pg;
-
-    use self::num_traits::{Signed, Zero, ToPrimitive};
-    use self::num_bigint::{Sign, BigInt, BigUint, ToBigInt};
-    use self::num_integer::Integer;
-    use self::bigdecimal::BigDecimal;
-
     use pg::data_types::PgNumeric;
     use types::{self, FromSql, ToSql, ToSqlOutput, IsNull};
 
-    type Digits = Vec<i16>;
-
-    fn bigdec_add_integer_part(digits: &mut Digits, absolute: &BigDecimal) -> i16 {
-        let mut weight = 0;
-        let ten_k = BigInt::from(10000);
-
-        let mut integer_part = absolute.to_bigint().expect("Can always take integer part of BigDecimal");
-
-        while ten_k <= integer_part {
-            weight += 1;
-            // digit is integer_part REM 10_000
-            let (div, digit) = integer_part.div_rem(&ten_k);
-            digits.push(digit.to_u16().expect("digit < 10000, but cannot fit in i16") as i16);
-            integer_part = div;
-        }
-        digits.push(integer_part.to_string().parse::<i16>().expect("digit < 10000, but cannot fit in i16"));
-
-        digits.reverse();
-
-        weight
-    }
-
-    fn bigdec_add_decimal_part(digits: &mut Digits, absolute: &BigDecimal) -> u16 {
-        use std::str::FromStr;
-
-        let ten_k = BigDecimal::from_str("10000").expect("Could not parse into BigDecimal");
-
-        let decimal_part = absolute;
-        let mut decimal_part = decimal_part - absolute.with_scale(0);
-        // scale is the amount of digits to print. to_string() includes a "0.",
-        // that's why the -2 is there.
-        let scale = if decimal_part == Zero::zero() {
-            0
-        } else {
-            decimal_part.to_string().len() as u16 - 2
-        };
-
-        while decimal_part != BigDecimal::zero() {
-            decimal_part *= &ten_k;
-            let digit = decimal_part.to_bigint().expect("Can always take integer part of BigDecimal");
-
-            // This can be simplified when github.com/akubera/bigdecimal-rs/issues/13 gets
-            // solved; decimal_part -= &digit; should suffice by then.
-            decimal_part -= BigDecimal::new(digit.clone(), 0);
-            digits.push(digit.to_u16().expect("digit < 10000, but cannot fit in i16") as i16);
+    /// Iterator over the digits of a big uint in base 10k.
+    /// The digits will be returned in little endian order.
+    struct ToBase10000(Option<BigUint>);
+
+    impl Iterator for ToBase10000 {
+        type Item = i16;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            self.0.take().map(|v| {
+                let (div, rem) = v.div_rem(&BigUint::from(10000u16));
```

Comment:
If `Div<u16>` returned `u16` instead of `BigUint` I'd happily use that instead. It seems like all the scalar operations are just converting to the `num` type under the hood?